hack/allocate.sh

```diff
@@ -61,41 +65,117 @@ containerdConfigPatches:
   [plugins."io.containerd.grpc.v1.cri".registry.mirrors."localhost:5000"]
     endpoint = ["http://kind-registry:5000"]
 EOF
+  sleep 10
+  kubectl wait pod --for=condition=Ready -l '!job-name' -n kube-system --timeout=5m
+
 }
 
 serving() {
   echo "${em}② Knative Serving${me}"
-  kubectl apply --filename https://github.com/knative/serving/releases/download/$serving_version/serving-crds.yaml
-  sleep 2
-  curl -L -s https://github.com/knative/serving/releases/download/$serving_version/serving-core.yaml | yq 'del(.spec.template.spec.containers[]?.resources)' -y | yq 'del(.metadata.annotations."knative.dev/example-checksum")' -y | kubectl apply -f -
+
+  kubectl apply --filename https://github.com/knative/serving/releases/download/$knative_version/serving-crds.yaml
+
+  sleep 5
+  kubectl wait --for=condition=Established --all crd --timeout=5m
+
+  curl -L -s https://github.com/knative/serving/releases/download/$knative_version/serving-core.yaml | yq 'del(.spec.template.spec.containers[]?.resources)' -y | yq 'del(.metadata.annotations."knative.dev/example-checksum")' -y | kubectl apply -f -
+
+
+  sleep 5
+  kubectl wait pod --for=condition=Ready -l '!job-name' -n knative-serving --timeout=5m
+
+  kubectl get pod -A
 }
 
-eventing() {
-  echo "${em}③ Knative Eventing${me}"
-  kubectl apply --filename https://github.com/knative/eventing/releases/download/$eventing_version/eventing-crds.yaml
-  sleep 2
-  curl -L -s https://github.com/knative/eventing/releases/download/$eventing_version/eventing-core.yaml | yq 'del(.spec.template.spec.containers[]?.resources)' -y | yq 'del(.metadata.annotations."knative.dev/example-checksum")' -y | kubectl apply -f -
-  curl -L -s https://github.com/knative/eventing/releases/download/$eventing_version/in-memory-channel.yaml | yq 'del(.spec.template.spec.containers[]?.resources)' -y | yq 'del(.metadata.annotations."knative.dev/example-checksum")' -y | kubectl apply -f -
-  curl -L -s https://github.com/knative/eventing/releases/download/$eventing_version/mt-channel-broker.yaml | yq 'del(.spec.template.spec.containers[]?.resources)' -y | yq 'del(.metadata.annotations."knative.dev/example-checksum")' -y | kubectl apply -f -
+dns() {
+  echo "${em}③ DNS${me}"
+  n=0
+  until [ $n -ge 10 ]; do
```

Comment:
Why are you looping here instead of looking for some kind of ready state? What happens when you try 10 times and they all fail. It doesn't seem like there is an explicit failure mode here.

Reply:
Thanks @lance and I share your general unease.  This is precisely what I was referring to in our call when I mentioned general flakiness.  This is one of the commands which I struggled with because it appears to have no condition upon which can be waited. The webhook process, which is supposedly ready (see wait from earlier step) fails with an IO timeout in low resource scenarios, but is considered ready.  Therefore, it seems there is simply a racing condition here and I have no choice but to simply try a few times and, as you say, let it fall through to failing on a later step if unavailable.

I agree this is ugly when compared with an explicit condition wait, and I am open to suggestions on how to improve it.  However, as you mention, if it fails 10 times (I never saw it go beyond 4), it will fail on the next kubectl call immediately.

Reply:
The two retry loops now fail-fast if `n` is reached.  Thanks @lance